### ISSUE TITLE
Pre- and post- deploy tasks in pipelines

### ISSUE THREAD

**Issue by dennisjbell** [genesis-community/genesis#427]
Add the ability to run scripts pre- and post- deployment tasks. This can be used to automatically create and close CR tickets, notify customers, etc...

Details remaining to be determined:

Is this only for deployments that are not automatically triggered?
Does each deployment task need separate inputs, or is the combination of Concourse env vars and the deployment task inputs sufficient to customize the scripts behaviour for all cases (prod vs non-prod, for example)?
Would the script live in the pipeline yml file, or would it be added to the deployment repo? There are pro/cons to both approaches.

**Comment by dennisjbell** [comment]
Rather than putting this in pipelines, meaning that manual deploys will have to manually reproduce the pre- and post-steps, probably introducing human error, lets keep this all within genesis repo and handle it git-ops-esque style.

Plausible implementation:

1) Add support for pre- and post-deploy scripts to each environment.

```
genesis:
  actions:
    pre-deploy:
      - addon: upload-stemcells
      - script: my_script
      - script: another_script
        args: [with,args,in,list]
    post-deploy
      - script: my_script
        args: [undo]
      - script: email_me
        args: [$USER_EMAIL]

 ```

Notes:
- scripts can be listed so they can be composable and reused
- scripts can be given arguments (including ENV VARS)
- scripts will also be passed environment variables describing the environment and in the case of post-deploy, the status of the action (pass/fail)
- Deploy options, including dry-run, is passed to the scripts so they can decide how to handle whether to run or not, or something in between
- Need to trap INT/TERM signals to ensure post-deploy hooks are called

**Comment by rob-greene** [comment]
If the `pre-deploy` or `post-deploy` scripts require access to Vault, is it "safe" to assume that `safe` is both available and authenticated? Or will there be some other mechanism around that?

**Comment by dennisjbell** [comment]
@rob-greene safe will already be authenticated, and its connection configuration will be part of the environment variables available to the script, so safe will 'just work'

**Comment by dennisjbell** [comment]
Released as part of v2.8.6
